src/lib/api-keys.ts

```diff
@@ -0,0 +1,290 @@
+ */
+const scopeChecks: ScopeCheckCollection = {
+	global: () => true,
+	app: (resources, { appId }) =>
+		resources.apps != null && resources.apps.includes(appId),
+};
+
+export function serialiseScopes(scopes: Scope[]): string {
+	return JSON.stringify(scopes);
+}
+
+export function deserialiseScopes(json: string): Scope[] {
+	return JSON.parse(json);
+}
+
+export function isScoped(
+	resources: Partial<ScopedResources>,
+	scopes: Scope[],
+): boolean {
+	return (
+		scopes.filter((scope) =>
+			scopeChecks[scope.type](resources, (scope as unknown) as any),
+		).length >= 1
+	);
+}
+
+export type AuthorizedRequest = express.Request & {
+	auth: {
+		isScoped: (resources: Partial<ScopedResources>) => boolean;
+		apiKey: string;
+		scopes: Scope[];
+	};
+};
+export type AuthorizedRequestHandler = (
+	req: AuthorizedRequest,
+	res: express.Response,
+	next: express.NextFunction,
+) => void;
+
+// empty until populated in `initialized`
+export let cloudApiKey: string = '';
+
+// should be called before trying to use this singleton
+export const initialized = (async () => {
+	await db.initialized;
+
+	// make sure we have an API key which the cloud will use to call us
+	await generateCloudKey();
+})();
+
+/**
+ * This middleware will extract an API key used to make a call, and then expand it out to provide
+ * access to the scopes it has. The `req` will be updated to include this `auth` data.
+ *
+ * E.g. `req.auth.scopes: []`
+ *
+ * @param req
+ * @param res
+ * @param next
+ */
+export const authMiddleware: AuthorizedRequestHandler = async (
+	req,
+	res,
+	next,
+) => {
+	// grab the API key used for the request
+	const apiKey = getApiKeyFromRequest(req) ?? '';
+
+	// store the key in the request, and an empty scopes array to populate after resolving the key scopes
+	req.auth = {
+		apiKey,
+		scopes: [],
+		isScoped: () => false,
+	};
+
+	try {
+		const conf = await config.getMany(['localMode', 'unmanaged', 'osVariant']);
+
+		// we only need to check the API key if a) unmanaged and on a production image, or b) managed and not in local mode
+		const needsAuth = conf.unmanaged
+			? conf.osVariant === 'prod'
+			: !conf.localMode;
+
+		// no need to authenticate, shortcut
+		if (!needsAuth) {
+			return next();
+		}
+
+		// if we have a key, find the scopes and add them to the request
+		if (apiKey && apiKey !== '') {
+			await initialized;
+			const scopes = await getScopesForKey(apiKey);
+
+			if (scopes != null) {
+				// keep the scopes for later incase they're desired
+				req.auth.scopes.push(...scopes);
+
+				// which resources are scoped...
+				req.auth.isScoped = (resources) => isScoped(resources, req.auth.scopes);
+
+				return next();
+			}
+		}
+
+		// we do not have a valid key...
+		return res.sendStatus(401);
+	} catch (err) {
+		console.error(err);
+		res.status(503).send(`Unexpected error: ${err}`);
+	}
+};
+
+function isEqualScope(a: Scope, b: Scope): boolean {
+	return _.isEqual(a, b);
```

Comment:
`Scope` types are (de)serialised to JSON so have to be flat objects, which makes this possible. Might be an issue if a scope becomes more complex.